compiler/passes/convert-typed-uast.cpp

```diff
@@ -214,11 +245,84 @@ struct TConverter final : UastConverter {
     // TODO: This won't work for variables declared within a 'param' loop!
     std::unordered_map<ID, Symbol*> localSyms;
 
-    // The current insertion point for expressions.
-    AList* aList = nullptr;
+    // The current AST list, its owner, and if it is for a statement.
+    std::vector<std::tuple<AList*, Expr*, bool>> aListStack;
+
+    // To fetch the previous statement children list.
+    std::vector<size_t> stmtListIdxStack;
 
-    // Used to determine if we are in a USER context or not.
+    // Used to quickly determine if we are in a user context or not.
     ModTag topLevelModTag = MOD_USER;
+
+    /// ------------------ ///
+    /// aListStack helpers ///
+    /// ------------------ ///
+
+    void pushAList(AList* alist, Expr* owner, bool isStmtList) {
+      int idx = (int) aListStack.size();
+      aListStack.push_back({alist, owner, isStmtList});
+      if (isStmtList) {
+        stmtListIdxStack.push_back(idx);
+      }
+    }
+
+    Expr* popAList() {
+      INT_ASSERT(aListStack.size() > 0);
+
+      Expr* ret = nullptr;
+      auto [alist, owner, isStmtList] = aListStack.back();
+      aListStack.pop_back();
+      std::ignore = alist;
+      ret = owner;
+
+      if (isStmtList) {
+        INT_ASSERT(!stmtListIdxStack.empty());
+        stmtListIdxStack.pop_back();
+      }
+
+      return ret;
+    }
+
+    std::tuple<AList*, Expr*, bool>& fetchListStackEntry(bool isStmtList) {
```

Comment:
let's rename isStmtList to getStatement or something as it's not asserting about the list it's indicating what to get

Reply:
How about `bool fetchStmtList`? I think we have the same idea. Thanks!